Details element polyfill

rmcwhae opened the pull request:
Adds a polyfill for HTML `<details>` elements (Issue https://github.com/Financial-Times/polyfill-library/issues/221).

The polyfill is taken from https://github.com/javan/details-element-polyfill. Tests here are ported versions of those found at https://github.com/javan/details-element-polyfill/blob/main/test/index.js.

Question: I pasted https://github.com/javan/details-element-polyfill/blob/main/dist/details-element-polyfill.js into `polyfill.js` here—is this the correct way of including said polyfill? The other approach, seen in some of the other polyfills, would be to `npm i details-element-polyfill` and add the following lines to `config.toml` (not sure if I should then remove the content in `polyfill.js` to avoid having two copies of the polyfill code):
```
[install]
module = "details-element-polyfill"
paths = [ "dist/details-element-polyfill.js" ]
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Financial-Times/polyfill-library?pullRequest=1102) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Financial-Times/polyfill-library?pullRequest=1102) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Financial-Times/polyfill-library?pullRequest=1102) it.</sub>

JakeChampion commented:
Hi @rmcwhae - thanks for the contribution!

> Question: I pasted https://github.com/javan/details-element-polyfill/blob/main/dist/details-element-polyfill.js into polyfill.js here—is this the correct way of including said polyfill? The other approach, seen in some of the other polyfills, would be to npm i details-element-polyfill and add the following lines to config.toml (not sure if I should then remove the content in polyfill.js to avoid having two copies of the polyfill code)

If the code pasted into the polyfill.js file in this project is exactly the same as the code in javan/details-element-polyfill then it is preferred to use the `config.toml` approach as we can more easily update to newer versions of the polyfill if any updates are released ☺️

JakeChampion commented on polyfills/HTMLDetailsElement/detect.js:
```diff
@@ -0,0 +1 @@
+"HTMLDetailsElement " in self;
```
This file is used as a predicate for an if statement, because of that it needs to be an expression, which means it can't have a semicolon at the end

```suggestion
"HTMLDetailsElement " in self
```

JakeChampion commented on polyfills/HTMLDetailsElement/polyfill.js:
```diff
@@ -0,0 +1,244 @@
+	var styles =
+		'\ndetails, summary {\n  display: block;\n}\ndetails:not([open]) > *:not(summary) {\n  display: none;\n}\nsummary::before {\n  content: "►";\n  padding-right: 0.3rem;\n  font-size: 0.6rem;\n  cursor: default;\n}\n[open] > summary::before {\n  content: "▼";\n}\n';
+	var _ref = [],
+		forEach = _ref.forEach,
```
Array.prototype.forEach will need to be added to the config.toml.dependencies array

rmcwhae replied:
Updated

JakeChampion commented on polyfills/HTMLDetailsElement/polyfill.js:
```diff
@@ -0,0 +1,244 @@
+		var prototype = document.createElement("details").constructor.prototype;
+		var setAttribute = prototype.setAttribute,
+			removeAttribute = prototype.removeAttribute;
+		var open = Object.getOwnPropertyDescriptor(prototype, "open");
```
Object.getOwnPropertyDescriptor will need to be added to the config.toml.dependencies array

rmcwhae replied:
Updated

JakeChampion commented on polyfills/HTMLDetailsElement/polyfill.js:
```diff
@@ -0,0 +1,244 @@
+		var setAttribute = prototype.setAttribute,
+			removeAttribute = prototype.removeAttribute;
+		var open = Object.getOwnPropertyDescriptor(prototype, "open");
+		Object.defineProperties(prototype, {
```
Object.defineProperties will need to be added to the config.toml.dependencies array

rmcwhae replied:
Updated

JakeChampion commented on polyfills/HTMLDetailsElement/polyfill.js:
```diff
@@ -0,0 +1,244 @@
+				childList: true
+			});
+		} else {
+			document.addEventListener("DOMNodeInserted", function (event) {
```
`Event` will need to be added to the config.toml.dependencies array as not all browsers we support have document.addEventListener

rmcwhae replied:
Updated

JakeChampion commented on polyfills/HTMLDetailsElement/polyfill.js:
```diff
@@ -0,0 +1,244 @@
+	}
+	function findClosestElementWithTagName(element, tagName) {
+		if (typeof element.closest == "function") {
+			return element.closest(tagName);
```
`Element.prototype.closest` could be added to the config.toml.dependencies array instead of using the inlined polyfill in the else block

rmcwhae replied:
Added

rmcwhae commented:
Thanks for the quick feedback! I have added the dependencies to `config.toml`.

> If the code pasted into the polyfill.js file in this project is exactly the same as the code in javan/details-element-polyfill then it is preferred to use the `config.toml` approach as we can more easily update to newer versions of the polyfill if any updates are released ☺️

Yes the code is exactly the same, so let’s go with `[install]` in `config.toml`. I added `details-element-polyfill` as a devDependency in the root-level package.json. Two more quick questions:
1. Do I include `package-lock.json` in this PR (package installed using Node 12.22.1)
2. Do I keep the pasted code in `polyfill.js`? Now we have it in two places…

JakeChampion commented:
> * Do I include `package-lock.json` in this PR (package installed using Node 12.22.1)
> 
> * Do I keep the pasted code in `polyfill.js`? Now we have it in two places…

Yes, the `package-lock.json` should be included as it will have changed when installing the new dependency 👍 

You are right, we don't need the `polyfill.js` file now that we have the config.toml.install section defined, which means the file will be created when running `npm run build`. I've removed the file and added it to `.eslintignore` and `.gitignore` 👍

romainmenke commented:
@rmcwhae Thank you for this!

I went over the code and noticed this bit : 

```js
  function polyfillProperties() {
    var prototype = document.createElement("details").constructor.prototype;
```

The polyfill alters the `HTMLUnknownElement` prototype which will likely cause issues for some users.

```js
	it('does not affect HTMLUnknownElement proto', function () {
		var unknownProto = document.createElement("unknownwefvxcdfg").constructor.prototype;
		proclaim.ok(("undefined" === typeof Object.getOwnPropertyDescriptor(unknownProto, "open")));
	});
```

<img width="596" alt="Screenshot 2021-09-24 at 16 26 12" src="https://user-images.githubusercontent.com/11521496/134691071-af635df2-cf81-42ad-90bd-e0d9e4003471.png">

It also does not expose `HTMLDetailsElement` which is why it fails the feature detect.
This should be fixed upstream.

rmcwhae commented:
@romainmenke Agreed, hijacking the `HTMLUnknownElement` prototype is not the best practice. Otherwise, this `<details>` polyfill (https://github.com/javan/details-element-polyfill) does seem to offer a more complete implementation than others (e.g. https://github.com/rstacruz/details-polyfill/blob/master/index.js).

I won't have time to look at updating the polyfill library for a bit, so I might put this PR on hold, unless anyone else has any suggestions?
